[Lim Hong Wei, Jovon] iP

JovonLim opened the pull request:
# Duke 
> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes/#3_Productivity_is_being_able_to_do_things_that_you_were_never_able_to_do_before_-_Franz_Kafka))

Duke frees your mind of having to remember things you need to do. It's
- Text-based
- Easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is:
1. Download the jar file from [here](https://github.com/JovonLim/ip/releases/tag/A-Jar)
1. Copy it into an empty folder
1. Open a command window in that folder and run `java -jar Duke.jar`
1. Add your tasks
1. Let it manage your tasks for you :wink:

And it is **FREE!** 😮 

Features:
- [X] Managing tasks
- [ ] Reminders (coming soon)

---
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Duke {
    public static void main(String[] args) {
        new Duke(System.getProperty("user.home")).run();
    }
}
```

Commit:
Add Gradle support

JovonLim committed:
Level 1

JovonLim committed:
Added add + list feature

JovonLim committed:
Added task class and ability to mark task as done

JovonLim committed:
Tasks are split into different categories

JovonLim committed:
Added testing for text and additional msgs for errors

JovonLim committed:
Created exception class for handling errors

JovonLim committed:
Added ability to delete tasks

JovonLim committed:
Reformat switch case

JovonLim committed:
Refactor with enum

JovonLim committed:
Implement save in hard disk

JovonLim committed:
Add Dates and Times to tasks

JovonLim committed:
Adjust text-UI tests for Level-7

JovonLim committed:
Merge branch 'branch-Level-7'

JovonLim committed:
Adjusted expected.txt

JovonLim committed:
Fix merge conflict of branch-Level-8 to master

JovonLim committed:
Add more OOP

JovonLim committed:
Organize classes into packages

JovonLim committed:
Merge remote-tracking branch 'origin/add-gradle-support'

JovonLim committed:
Fix mark and unmark bug

JovonLim committed:
Set up gradle with checkstyle-support

JovonLim committed:
Fix checkstyle for some classes

JovonLim committed:
Add JUnit tests

JovonLim committed:
Remove emote from msg

JovonLim committed:
Add JavaDocs to all public classes

JovonLim committed:
Adhere to coding standard

JovonLim committed:
Add find command

JovonLim committed:
Merge branch 'A-JavaDoc'

JovonLim committed:
Merge branch 'branch-A-CodingStandard'

JovonLim committed:
Merge branch 'branch-Level-9'

JovonLim committed:
Fix checkstyles from merge

JovonLim committed:
Write to file during taskfile altering commands rather than at exit

JovonLim committed:
Update javadocs for affected commands

JovonLim committed:
Update task

JovonLim committed:
Ensure no checkstyle errors and added more javadocs

JovonLim committed:
Add a GUI to Duke using JavaFX

JovonLim committed:
Apply Varargs to find command

JovonLim committed:
Merge branch 'branch-A-Checkstyle'

JovonLim committed:
Merge branch 'branch-Level-10'

JovonLim committed:
Merge branch 'branch-A-Varargs'

JovonLim committed:
Fix tasktype error and add Javadoc for public method in Duke

JovonLim committed:
Update javadocs

WingKei67 commented on src/main/java/duke/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,32 @@
+     *
+     * @param list The taskList of Duke.
+     * @param storage The fileStorage of Duke.
+     */
```
Would it be better if you can add @return tag?

WingKei67 commented on src/main/java/duke/command/ExitCommand.java:
```diff
@@ -0,0 +1,22 @@
+     */
+    @Override
+    public String execute(TaskList list, FileStorage storage) {
+        return "Farewell, till we meet again";
```
Would it be better if showing messages is handled by UI class?

WingKei67 commented on src/main/java/duke/gui/DialogBox.java:
```diff
@@ -0,0 +1,74 @@
+import javafx.scene.control.Label;
+import javafx.scene.image.Image;
+import javafx.scene.image.ImageView;
+import javafx.scene.layout.HBox;
```
I really appreciated how you organize your imported packages by order.

WingKei67 commented on src/main/java/duke/exception/DukeFileException.java:
```diff
@@ -0,0 +1,10 @@
+
+/**
+ * Creates an exception that is thrown when the file storage fails to read or write.
+ */
```
Would it be better if you can add @throws tag?

WingKei67 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,277 @@
+     * @return An arraylist of strings to be written.
+     */
+    public ArrayList<String> writeFileContents(ArrayList<Task> taskList) {
+        ArrayList<String> list = new ArrayList<>();
```
Would it be better if the naming could be more specific such as tasks?

WingKei67 commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,121 @@
+                    tmp.add(task);
+                }
+            }
+        }
```
I really appreciated how you wrapped the loop body by curly brackets.

WingKei67 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,83 @@
+        } else {
+            return 0;
+        }
+    }
```
Would it be more succinct if you use ternary operator?

WingKei67 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,83 @@
+     * @param date The given LocalDate.
+     * @return The boolean value of whether the date matches.
+     */
+    public abstract boolean isDateEqual(LocalDate date);
```
I like how you name boolean variables.

JovonLim committed:
Fix javadocs, and readd Ui class to abstract the message in command classes

JovonLim committed:
Add assertions into parser and duke class

Parser and Duke classes have underlying assumptions in their methods.

Assertions will allow us to document them and ensure these assumptions hold.

Add assertions to switch cases with no default case and to check that
duke commands and responses are non-null and non-empty respectively.

Assertions are a quick and efficient ways to find bugs.

JovonLim committed:
Adhere to given Code Quality guidelines

Current classes contain some magic literals, long methods and arrowhead
code.

These decrease readability of code.

Refactoring is done to split long methods into shorter methods and
remove arrowhead code. Magic literals have been rewritten to class
fields for error messages and duke's response.

By splitting longer methods into shorter methods, it allows the reader
to have an easier time tracking what the code does.

JovonLim committed:
Add streams to tasklist methods

Find and Due commands are done by iterating through for loops.

To allow these commands to be more flexible in the future for sorting
etc.

Utilize streams for their filter method and collect them into a list.

Easier readability as well as being able to add more methods to the
stream if needed.

JovonLim committed:
Merge pull request #1 from JovonLim/branch-A-Assertions

Add assertions into parser and duke class

JovonLim committed:
Merge branch 'master' into branch-A-CodeQuality

JovonLim committed:
Merge branch 'master' into branch-A-Streams

JovonLim committed:
Merge pull request #2 from JovonLim/branch-A-CodeQuality

Adhere to given Code Quality guidelines

JovonLim committed:
Merge branch 'master' into branch-A-Streams

JovonLim committed:
Merge pull request #3 from JovonLim/branch-A-Streams

Add streams to tasklist methods

JovonLim committed:
Set up CI

JovonLim committed:
Add priority to tasks

JovonLim committed:
Remove test class and fixed coding standard violations

JovonLim committed:
Merge pull request #4 from JovonLim/branch-C-Priority

Add priority to tasks

JovonLim committed:
Fix priority command

NicholasPatrick commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,127 @@
+/**
+ * A component of the chatBot Duke that returns his response as a string.
+ */
+public class Ui {
```
Have you considered making all the methods in this class static, considering `Ui` doesn't have any non-static attributes?

NicholasPatrick commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,361 @@
+    }
+
+
+    private String appendDescription(String ... components) {
+        StringBuilder str = new StringBuilder();
+        for (String component : components) {
+            str.append(component).append(" ");
+        }
+        return str.toString();
+    }
```
Did you know that `String.join` exists and does exactly this?

NicholasPatrick commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,72 @@
+public class Deadline extends Task {
+    protected LocalDate by;
+    protected LocalTime time;
+    protected DateTimeFormatter timeFormatter = DateTimeFormatter.ofPattern("HH:mm");
+    protected DateTimeFormatter dateFormatter = DateTimeFormatter.ofPattern("MMM d yyyy");
```
Perhaps you can make these values `final`?

NicholasPatrick commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,72 @@
+ * Represents a task that is happening at a date, and its duration.
+ */
+public class Event extends Task {
+    protected LocalDate at;
```
This variable name could perhaps be made more descriptive, especially considering the fact that when specifying a date, the grammatically correct preposition is "on", not "at".
"On the first of August," not "at the first of August."

NicholasPatrick reviewed (commented):
Overall, a great use of OOP 👍. Just a few nits here and there.

JovonLim committed:
Fix code quality issue and todo bug

JovonLim committed:
Remove unused import

JovonLim committed:
Update GUI

JovonLim committed:
Update README.md

JovonLim committed:
Add screenshot to docs

JovonLim committed:
Merge branch 'master' of https://github.com/JovonLim/ip

JovonLim committed:
Simplify priority encoding

JovonLim committed:
Update JUnit Tests

JovonLim committed:
Improve clarity of error message for priority command

JovonLim committed:
Change Gigachad response

JovonLim committed:
Add parserTest

JovonLim committed:
Update parser method

JovonLim committed:
Update Gui wrapping of text

JovonLim committed:
Update GUI

JovonLim committed:
Update README.md

JovonLim committed:
Adjust encoding sequence for tasks

JovonLim committed:
Merge branch 'master' of https://github.com/JovonLim/ip

JovonLim committed:
Add help command

JovonLim committed:
Update README.md

JovonLim committed:
Update README.md

JovonLim committed:
Update README.md

JovonLim committed:
Update query and remove dead code

JovonLim committed:
Fix coding standard violation

JovonLim committed:
Update README.md

JovonLim committed:
Update README.md

JovonLim committed:
Update README.md

JovonLim committed:
Update README.md

JovonLim committed:
Update README.md
